test/pickfirst_leaf_test.go

```diff
@@ -0,0 +1,607 @@
+	"google.golang.org/grpc/internal/testutils/pickfirst"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+	"google.golang.org/grpc/status"
+
+	testgrpc "google.golang.org/grpc/interop/grpc_testing"
+	testpb "google.golang.org/grpc/interop/grpc_testing"
+)
+
+var stateStoringServiceConfig = fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, stateStoringBalancerName)
+
+const stateStoringBalancerName = "state_storing"
+
+// setupPickFirstLeaf performs steps required for pick_first tests. It starts a
+// bunch of backends exporting the TestService, creates a ClientConn to them
+// with service config specifying the use of the pick_first LB policy.
+func setupPickFirstLeaf(t *testing.T, backendCount int, opts ...grpc.DialOption) (*grpc.ClientConn, *manual.Resolver, []*stubserver.StubServer) {
+	t.Helper()
+
+	r := manual.NewBuilderWithScheme("whatever")
+
+	backends := make([]*stubserver.StubServer, backendCount)
+	addrs := make([]resolver.Address, backendCount)
+	for i := 0; i < backendCount; i++ {
+		backend := &stubserver.StubServer{
+			EmptyCallF: func(_ context.Context, _ *testpb.Empty) (*testpb.Empty, error) {
+				return &testpb.Empty{}, nil
+			},
+		}
+		if err := backend.StartServer(); err != nil {
+			t.Fatalf("Failed to start backend: %v", err)
+		}
+		t.Logf("Started TestService backend at: %q", backend.Address)
+		t.Cleanup(func() { backend.Stop() })
+
+		backends[i] = backend
+		addrs[i] = resolver.Address{Addr: backend.Address}
+	}
+
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithDefaultServiceConfig(stateStoringServiceConfig),
+	}
+	dopts = append(dopts, opts...)
+	cc, err := grpc.NewClient(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// At this point, the resolver has not returned any addresses to the channel.
+	// This RPC must block until the context expires.
+	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer sCancel()
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(sCtx, &testpb.Empty{}); status.Code(err) != codes.DeadlineExceeded {
+		t.Fatalf("EmptyCall() = %s, want %s", status.Code(err), codes.DeadlineExceeded)
+	}
+	return cc, r, backends
+}
+
+type scStateExpectation struct {
+	State     connectivity.State
+	ServerIdx int
+}
+
+func scStateExpectationToScState(in []scStateExpectation, serverAddrs []resolver.Address) []scState {
+	out := []scState{}
+	for _, exp := range in {
+		out = append(out, scState{
+			State: exp.State,
+			Addrs: []resolver.Address{serverAddrs[exp.ServerIdx]},
+		})
+	}
+	return out
+
+}
+
+// TestPickFirstLeaf_ResolverUpdate tests the behaviour of the new pick first
+// policy when servers are brought down and resolver updates are received.
+func (s) TestPickFirstLeaf_ResolverUpdate(t *testing.T) {
```

Comment:
I appreciate your effort in making this a table driven test. But the table has way too many fields I feel and the test is very hard to read because of that. Would you mind breaking them up and writing smaller helpers for shared functionailty, thereby reducing the size of each test, but making each test self-contained and more readable.